src/complete.cpp

```diff
@@ -1454,20 +1445,140 @@ void completer_t::mark_completions_duplicating_arguments(const wcstring &prefix,
 }
 
 /// Return the index of an argument from \p args containing the position \p pos, or none if none.
-static maybe_t<size_t> find_argument_containing_position(const arg_list_t &args, size_t pos) {
+static maybe_t<size_t> find_argument_containing_position(
+    const std::vector<const parse_node_t *> &args, size_t pos) {
     size_t idx = 0;
     for (const auto &arg : args) {
-        if (arg.location_in_or_at_end_of_source_range(pos)) {
+        if (arg->location_in_or_at_end_of_source_range(pos)) {
             return idx;
         }
         idx++;
     }
     return none();
 }
 
+/// This takes a parse tree and a source offset and computes a vector of
+/// tokens, similar to the output of `commandline -op`.
+///
+/// This assumes that the tree is computed off of a process range
+/// as returend by parse_util_process_extent, which means there
+/// are no ; & | && || in the AST, which simplifies the traversal.
+///
+/// The returned nodes are string nodes (keywords for builtins like "if"
+/// and "not" and the command string), and a number of argument nodes at the end.
+static std::vector<const parse_node_t *> collect_tokens_of_command_at_location(
```

Comment:
You can use `tnode_t::descendants()` for this:

    auto strings = statement.descendants<grammar::tok_string>();

This gives you all the string nodes under a given node (recursively).

Reply:
Yes I didn't realize the same can be done with one `tnode_t::descendants()`.
But what I am doing now is even simpler, I'm just using the tokenizer because with this change the completion does not need to use the AST, it just needs the string tokens in a process. So this seems good enough for now.